src/modules-load/modules-load.c

```diff
@@ -150,35 +407,69 @@ static int run(int argc, char *argv[]) {
-                        RET_GATHER(r, apply_file(ctx, argv[i], false));
-
+                for (int i = optind; i < argc; i++) {
+                        r = apply_file_from_path(argv[i], &module_set);
+                        if (r < 0)
+                                RET_GATHER(ret, r);
+                }
         } else {
-                _cleanup_strv_free_ char **files = NULL;
+                ConfFile **files = NULL;
+                size_t n_files = 0;
+
+                CLEANUP_ARRAY(files, n_files, conf_file_free_many);
 
                 STRV_FOREACH(i, arg_proc_cmdline_modules) {
-                        k = module_load_and_warn(ctx, *i, true);
-                        if (k == -ENOENT)
-                                continue;
-                        RET_GATHER(r, k);
+                        r = modules_list_append_dup(&module_set, *i);
+                        if (r < 0)
+                                RET_GATHER(ret, log_error_errno(r, "Failed to add module '%s': %m", *i));
                 }
 
-                k = conf_files_list_nulstr(&files, ".conf", NULL, 0, conf_file_dirs);
-                if (k < 0)
-                        return log_error_errno(k, "Failed to enumerate modules-load.d files: %m");
+                r = conf_files_list_nulstr_full(".conf", /* root = */ NULL,
+                                                CONF_FILES_REGULAR | CONF_FILES_FILTER_MASKED,
+                                                conf_file_dirs, &files, &n_files);
+                if (r < 0)
+                        RET_GATHER(ret, log_error_errno(r, "Failed to enumerate modules-load.d files: %m"));
+                else
+                        FOREACH_ARRAY(cf, files, n_files)
+                                RET_GATHER(ret, apply_conf_file(*cf, &module_set));
+        }
+
+        n_threads = determine_num_worker_threads((size_t) ordered_set_size(module_set));
 
-                STRV_FOREACH(fn, files)
-                        RET_GATHER(r, apply_file(ctx, *fn, true));
+        /* If no additional thread is required, there is no need to create the
+         * thread pool or the mean to communicate with its members. */
+        if (n_threads == 0) {
+                log_debug("Single-threaded probe");
+                return RET_GATHER(ret, do_direct_probe(module_set));
         }
 
-        return r;
+        /* Create a socketpair for communication with probe workers */
+        r = RET_NERRNO(socketpair(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0, pair));
+        if (r < 0)
+                return log_error_errno(r, "Failed to create socket pair: %m");
+
+        /* Create threads, which will then wait for modules to probe. */
+        log_info("Using %zu probe threads", n_threads + 1);
+        r = create_worker_threads(n_threads, FD_TO_PTR(pair[1]), &threads, &n_threads);
+        if (r < 0)
+                log_warning("Failed to create probe threads, continuing as single-threaded probe");
```

Comment:
even  if you do not use %m in a log message, do pass the error code this is about, if you have one (i.e. use log_warning_errno() here), since this generally goes into structured logging